src/common.coffee

```diff
@@ -36,19 +54,88 @@ isDescVal = (v) ->
     when -1, '-1', 'desc', 'descending' then true
     else false
 
+
+isRaw = (q) ->
+  TYPES.RAW is q.type
+
+isRegularWrite = (q) ->
+  switch q.type
+    when TYPES.INSERT, TYPES.UPDATE, TYPES.REMOVE
+      true
+    else
+      false
+
+isRawWrite = (q) ->
+  return false unless isRaw q
+  switch q.action_type
+    when ACTION_TYPES.INSERT, ACTION_TYPES.UPDATE, ACTION_TYPES.REMOVE
+      true
+    else
+      false
+
+isWrite = (q) ->
+  isRegularWrite(q) or isRawWrite(q)
+
+# classification specifies if it's the regular,
+# raw or either type. Default is 'either'
+validateClassification = (classification) ->
+  return 'either' unless 'string' is typeof classification
+  classification = classification.trim().toLowerCase()
+  switch classification
+    when 'regular', 'raw'
+      classification
+    else
+      'either'
+
+checkRegularType = (q, checkType) ->
+  TYPES[checkType] is q.type
+
+checkRawType = (q, checkType) ->
+  (TYPES.RAW is q.type) and (ACTION_TYPES[checkType] is q.action_type)
+
+commonCheckType = (q, checkType, classification = 'either') ->
+  classification = validateClassification classification
+  switch classification
+    when 'regular'
+      checkRegularType q, checkType
+    when 'raw'
+      checkRawType q, checkType
+    when 'either'
+      (checkRegularType q, checkType) or (checkRawType q, checkType)
+
+isInsert = (q, classification = 'either') ->
+  commonCheckType q, 'INSERT', classification
+
+isUpdate = (q, classification = 'either') ->
+  q.upsert isnt true and commonCheckType q, 'UPDATE', classification
+
+isUpsert = (q, classification = 'either') ->
+  q.upsert is true and commonCheckType q, 'UPDATE', classification
+
+isRemove = (q, classification = 'either') ->
+  commonCheckType q, 'REMOVE', classification
+
+
 module.exports =
-  TYPES:
-    Q: 'Q'
-    AND: 'AND'
-    OR: 'OR'
-    RAW: 'RAW'
+  TYPES: TYPES
+  ACTION_TYPES: ACTION_TYPES
 
   preprocess: (q) ->
-    processed =
-      query: canonicalizeQuery q
-      options: canonicalizeOpts q
-      facet: q.facet
-    return processed
+    q = _.cloneDeep q
```

Comment:
I used to have a couple cloneDeeps in here but removed them for performance reasons. Is there a particular reason this is here? Are we mutating the query as we process it?

Can we follow the pattern from before, where we create a new object (`processed`) and write into that, instead of returning a modified clone?